### ISSUE TITLE
performance issues

### ISSUE THREAD

**Issue by marbemac** [jagi/meteor-astronomy#174]
Hi @jagi - uncovering more questions as I continue to transition our large app to astronomy. Really liking the library :), butttt now we're running into performance issues.

I've create a simple, isolated, and reproducible example at the repository below:
https://github.com/marbemac/astronomy-perf

Just pull it down, run it, and it will output the time taken to load the models via astronomy versus plain meteor. On my machine, I'm seeing ~550ms for astronomy, and around 3ms for meteor. The real pain here is that the entire page is blocked for those 550ms. We allow users to click through the list of elements, which causes a refresh of the models, and another 550ms freeze.

From what I can tell, adding nested models significantly increases that load time. For example, if you comment out the two big nested models (which have nested models of their own inside):

https://github.com/stoplightio/astronomy-perf/blob/master/model.js#L353
https://github.com/stoplightio/astronomy-perf/blob/master/model.js#L361

The block time drops to ~12ms for astronomy, which is awesome.

Do you think there's any low hanging fruit here performance-wise?

**Comment by jagi** [comment]
Hmm strange. I will take a look at it and try to find a bottleneck. Maybe it's because of recursion.

**Comment by jagi** [comment]
Ok I've found a problem. It's in the `ejson` package. It's very very slow. It's almost 8 times slower than standard JSON. I will try to limit its usage.

**Comment by jagi** [comment]
Unfortunately there is nothing I can do about it now. I'm using EJSON in many places. To get rid of it I would have to change the approach for storing the old values before modifications and it would require rewriting several modules.

I will have to think about not using EJSON in future Astronomy versions. It may be possible but will have some limitations. We will see. I saw that some people had already been complaining about EJSON being very slow.

For now, I can't suggest any solution for that... :( I will probably have to get rid of nested classes at all and change it to be more "plain" structure.

**Comment by marbemac** [comment]
Ah ok I see. So do you think most of the issue is due to EJSON use? If EJSON is ~8 times slower, does that mean that even if everything in astronomy was plain JSON, it would still take ~70ms to fetch/initialize ~30 of these models w nested classes (vs 3ms w/o astronomy)?

Imho trading some ease of use / features to alleviate this blocking would be well worth it for a future version. For now, I'll avoid using astronomy when fetching lists of models, and use it for our resource pages (editing one model at a time).

Thanks for taking a look at this so promptly btw.

**Comment by jagi** [comment]
It's hard to say what performance increase would I get when not using EJSON because it would not work the way it works right now. But we can estimate that making EJSON 8 times faster would make Astronomy approximately 8 times faster.

In your reproduction, on my computer, using Astronomy it takes around 900 ms to load data. After some changes I was able to get down to around 30 ms. But it cased a few features to be missing. I will definitely need to avoid using EJSON at all in the future. I have some ideas how to achieve that. For example I can postpone some EJSON calculation only to happen when a given feature is used. But it requires rethinking entire internal structure.

However, thanks for informing about this problem. It will help me develop Astronomy in the correct way. For now, I have a few ideas that will make it faster. I just have to bring back some feature from the beta version of Astronomy. I've chosen nested classes as a better way of structuring data and in fact it's a better way but it comes with a cost (because of EJSON). So I have come back to the previous solution for nesting data. However, some people will be happy with that because they were writing to me that they miss it from the beta version :).

What I can say. It's hard to make good decisions when developing library and not testing it in a big project :)

**Comment by jagi** [comment]
Good news! Currently, I'm working on performance improvement and I've managed to increase Astronomy speed. I've made it 10 times faster with a possibility to improve it even further but we will see what can be done.

There is also one bad news. It will appear in version 2.0 because I had to make some compatibility breaking changes. There won't be many changes in API. If it goes about nested fields it's gonna change from:

```js
fields: {
  address: {
    type: 'object',
    nested: 'Address'
  },
  phones: {
    type: 'array',
    nested: 'Phone'
  }
}
```

to:

```js
nested: {
  address: {
    count: 'one',
    class: 'Address'
  },
  phones: {
    count: 'many',
    class: 'Phone'
  }
}
```

There will some other changes and improvements as it will be the 2.0 version. However, it will take some time to implement. So you have to be patient.

**Comment by marbemac** [comment]
Awesome! Really happy to hear this :). Great work, excited to try it out when it's ready.

**Comment by jagi** [comment]
I will notify :) Stay tuned

**Comment by shwaydogg** [comment]
@marbemac was your performance slow down isolated to fetch's and building astro object?  I'm curious / confirming there was no issue with the subscription slowing down.

**Comment by jagi** [comment]
@shwaydogg it was definitely related with fetching not subscribing. I've checked it :). Current version of Astronomy 2.0 is very very fast.

Here are the times from the reproduction by @marbemac:
- Without Astronomy: **~ 3ms**
- Astronomy 1.0: **~ 500ms**
- Astronomy 2.0: **~ 22ms**

I think that this improvement will be noticeable :). There will be also more features :). Can wait to release it :). But it's still needs a little bit of work :)

**Comment by shwaydogg** [comment]
@jagi awesome, looking forward to it!  It seems like 90% off issues have been related to nested classes, so def don't mind the breaking changes there and think your new syntax above may be more readable!

**Comment by jagi** [comment]
Yes most posted issues are related to nested classes. They still gonna be available but in a little bit different way :). I've redesigned many parts of the package. I hope that this update will cause less issues :)

**Comment by marbemac** [comment]
Dude, badass. Can't wait to try it out. Any sort of timeline?

**Comment by jagi** [comment]
I hope it will be for Christmas but can't promise :)

**Comment by jagi** [comment]
I've made some performance improvements in 1.2.8 and 1.2.9. More improvements will be available in 2.0

**Comment by marbemac** [comment]
Nice work, down to ~30ms from ~500ms! :)
